csrc/abstract_tensor.h

```diff
@@ -332,10 +849,139 @@ struct AbstractTensor {
   //   row2  id4    id5
   // in another word, the return value will be an AbstractTensor:
   // [dim0={id0, id1, id4}, dim1={id2, id3, id5}].
-  AbstractTensor& addRow(AbstractTensor tensor);
+  AbstractTensorWithInfo& addRow(AbstractTensorWithInfo tensor) {
+    NVF_CHECK(
+        domain.size() == tensor.domain.size(),
+        "Can not add a new row with different number of domains.");
+    NVF_CHECK(
+        std::all_of(
+            domain.begin(),
+            domain.end(),
+            [](const AbstractId& aid) { return aid.is<std::vector>(); }),
+        "Can not add a new row to an AbstractTensor with non-vector domains.")
+
+    for (auto i : c10::irange(domain.size())) {
+      domain[i].as<std::vector>().emplace_back(std::move(tensor.domain[i]));
+    }
+    return *this;
+  }
 
   // Remove all the null elements.
-  AbstractTensor& strip();
+  AbstractTensorWithInfo& strip() {
+    AbstractTensorWithInfo result;
+    for (const auto& aid : domain) {
+      if (aid.hasValue()) {
+        result.pushBack(aid);
+      }
+    }
+    std::swap(result, *this);
+    return *this;
+  }
+};
+
+struct EmptyInfo {
+  static EmptyInfo merge(const EmptyInfo&, const EmptyInfo&) {
+    return {};
+  }
+
+  static std::pair<EmptyInfo, EmptyInfo> split(const EmptyInfo& a) {
+    return {{}, {}};
+  }
+
+  static std::pair<EmptyInfo, EmptyInfo> swizzle(
+      const EmptyInfo& a,
+      const EmptyInfo& b) {
+    return {{}, {}};
+  }
+
+  bool operator==(const EmptyInfo& t) const {
+    return true;
+  }
+};
+
+using AbstractTensor = AbstractTensorWithInfo<EmptyInfo>;
+
+//! This is a holds a comparable type of "tag". This version is strict in the
+//! sense that we check that the tags match whenever we do a merge instead of
+//! allowing mismatches.
```

Comment:
Where is the code that does the check:
> This version is strict in the sense that we check that the tags match whenever we do a merge instead of allowing mismatches.

Reply:
Error in the comment. It should say we just union the sets upon merge or swizzle. Will update.